authmanager/proxy/proxy_ws.go

```diff
@@ -0,0 +1,207 @@
+// Copyright 2021 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+package proxy
+
+// Heavily inspired by https://github.com/koding/websocketproxy
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+
+	"github.com/gorilla/websocket"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	maxWsConnections = 1
+)
+
+var (
+	// From https://github.com/mendersoftware/deviceconnect/blob/master/api/http/device.go
+	upgrader = websocket.Upgrader{
+		Subprotocols: []string{"protomsg/msgpack"},
+		CheckOrigin: func(r *http.Request) bool {
+			return true
+		},
+	}
+)
+
+func wsSchemeFromHttpScheme(httpScheme string) string {
+	switch httpScheme {
+	case "http":
+		return "ws"
+	case "https":
+		return "wss"
+	default:
+		return ""
+	}
+}
+
+func (pc *ProxyController) DoWsUpgrade(w http.ResponseWriter, r *http.Request) {
+	// Convert server request to client request and override r.URL
+	r.RequestURI = ""
+	r.Host = ""
+	r.URL.Scheme = wsSchemeFromHttpScheme(pc.conf.backend.Scheme)
+	r.URL.Host = pc.conf.backend.Host
+
+	requestHeader := http.Header{}
+	if origin := r.Header.Get("Origin"); origin != "" {
+		requestHeader.Add("Origin", origin)
+	}
+	for _, prot := range r.Header[http.CanonicalHeaderKey("Sec-WebSocket-Protocol")] {
+		requestHeader.Add("Sec-WebSocket-Protocol", prot)
+	}
+	for _, cookie := range r.Header[http.CanonicalHeaderKey("Cookie")] {
+		requestHeader.Add("Cookie", cookie)
+	}
+	if r.Host != "" {
+		requestHeader.Set("Host", r.Host)
+	}
+
+	requestHeader.Add("Authorization", r.Header.Get("Authorization"))
```

Comment:
I don't have a good answer. I wrote this filtering while learning about `websockets`. But now looking at it with perspective I should pass all of them, same as we do for every other API request.

I'll pass all and re-test `mender-connect`.

Reply:
I remember a bit more why selecting headers. The handshake coming from the client contains the websocket handshake headers like `Sec-Websocket-Key`, `Sec-Websocket-Version`, etc. which cannot be propagated to the backend. The proxy will call `Dial` and there create these headers again, so it would end up with duplicated (and different) headers.

In any case I got the spirit of your comment and instead of "opt in" few headers I will "opt out" the ones that are websockets handshake specific.